File: internal/handlers/db_viewer.go
package handlers

import (
	"context"
	"database/sql"
	"fmt"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
)

// TableInfo holds metadata about a database table
type TableInfo struct {
	Name     string
	RowCount int64
}

// ColumnInfo holds metadata about a table column
type ColumnInfo struct {
	Name     string
	Type     string
	Nullable bool
	Default  sql.NullString
}

// HandleDBViewer renders the database viewer page
// Route: GET /development/db
func (h *Handler) HandleDBViewer(c *gin.Context) {
	// Check if user is authenticated
	user, err := h.GetSessionUser(c)
	if err != nil || user == nil {
		// Redirect to sign-in page
		http.Redirect(c.Writer, c.Request, "/sign-in", http.StatusSeeOther)
		return
	}

	data := map[string]any{
		"Title":       "Database Viewer",
		"DBConnected": h.hasDB(),
		"User":        user,
	}

	if h.hasDB() {
		tables, err := h.getTableList(c.Request.Context())
		if err != nil {
			data["Error"] = fmt.Sprintf("Failed to get tables: %v", err)
		} else {
			data["Tables"] = tables
		}

		// Check if a table is selected via query param
		selectedTable := c.Query("table")
		if selectedTable != "" {
			data["SelectedTable"] = selectedTable
		}
	}

	h.render(c, "db_viewer", data)
}

// HandleDBTableContent returns the table content as an HTMX fragment
// Route: GET /hx/development/db/table/{name}
func (h *Handler) HandleDBTableContent(c *gin.Context) {
	if !h.hasDB() {
		http.Error(c.Writer, "Database not connected", http.StatusServiceUnavailable)
		return
	}

	tableName := c.Param("name")
	if tableName == "" {
		http.Error(c.Writer, "Table name is required", http.StatusBadRequest)
		return
	}

	// Parse pagination params
	offset, _ := strconv.Atoi(c.Query("offset"))
	limit, _ := strconv.Atoi(c.Query("limit"))
	if limit <= 0 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}

	data, err := h.getTableData(c.Request.Context(), tableName, offset, limit)
	if err != nil {
		http.Error(c.Writer, fmt.Sprintf("Failed to get table data: %v", err), http.StatusInternalServerError)
		return
	}

	h.renderPartial(c, "db_table_content", data)
}

// getTableList retrieves all user tables from the database
func (h *Handler) getTableList(ctx context.Context) ([]TableInfo, error) {
	var tableNames []string
	if err := h.db.WithContext(ctx).
		Table("information_schema.tables").
		Select("table_name").
		Where("table_schema = ? AND table_type = ?", "public", "BASE TABLE").
		Order("table_name").
		Scan(&tableNames).Error; err != nil {
		return nil, err
	}

	var tables []TableInfo
	for _, name := range tableNames {
		var colCount int64
		if err := h.db.WithContext(ctx).
			Table("information_schema.columns").
			Where("table_schema = ? AND table_name = ?", "public", name).
			Count(&colCount).Error; err != nil {
			return nil, err
		}

		var rowCount int64
		if err := h.db.WithContext(ctx).Table(name).Count(&rowCount).Error; err != nil {
			rowCount = 0
		}

		tables = append(tables, TableInfo{Name: name, RowCount: rowCount})
	}

	return tables, nil
}

// getTableData retrieves data and schema for a specific table
func (h *Handler) getTableData(ctx context.Context, tableName string, offset, limit int) (map[string]any, error) {
	data := map[string]any{
		"SelectedTable": tableName,
		"Offset":        offset,
		"Limit":         limit,
	}

	allowed, err := h.isTableAllowed(ctx, tableName)
	if err != nil {
		return nil, fmt.Errorf("failed to validate table: %w", err)
	}
	if !allowed {
		return nil, fmt.Errorf("invalid table name")
	}

	// Get column info
	columns, err := h.getColumnInfo(ctx, tableName)
	if err != nil {
		return nil, fmt.Errorf("failed to get columns: %w", err)
	}
	data["Columns"] = columns

	// Get total row count
	var totalRows int64
	if err := h.db.WithContext(ctx).Table(tableName).Count(&totalRows).Error; err != nil {
		return nil, fmt.Errorf("failed to count rows: %w", err)
	}
	data["TotalRows"] = totalRows

	// Get rows
	var rowMaps []map[string]any
	query := h.db.WithContext(ctx).Table(tableName).Limit(limit).Offset(offset)
	if len(columns) > 0 {
		query = query.Order(columns[0].Name)
	}
	if err := query.Find(&rowMaps).Error; err != nil {
		return nil, fmt.Errorf("failed to query rows: %w", err)
	}

	var rowData [][]any
	for _, row := range rowMaps {
		rowValues := make([]any, len(columns))
		for i, col := range columns {
			rowValues[i] = formatValue(row[col.Name])
		}
		rowData = append(rowData, rowValues)
	}
	data["Rows"] = rowData

	return data, nil
}

// getColumnInfo retrieves column metadata for a table
func (h *Handler) getColumnInfo(ctx context.Context, tableName string) ([]ColumnInfo, error) {
	var rawColumns []struct {
		Name     string
		Type     string
		Nullable string
		Default  sql.NullString
	}

	if err := h.db.WithContext(ctx).
		Table("information_schema.columns").
		Select("column_name as name, data_type as type, is_nullable as nullable, column_default as default").
		Where("table_schema = ? AND table_name = ?", "public", tableName).
		Order("ordinal_position").
		Scan(&rawColumns).Error; err != nil {
		return nil, err
	}

	columns := make([]ColumnInfo, 0, len(rawColumns))
	for _, col := range rawColumns {
		columns = append(columns, ColumnInfo{
			Name:     col.Name,
			Type:     col.Type,
			Nullable: col.Nullable == "YES",
			Default:  col.Default,
		})
	}

	return columns, nil
}

func (h *Handler) isTableAllowed(ctx context.Context, tableName string) (bool, error) {
	var count int64
	if err := h.db.WithContext(ctx).
		Table("information_schema.tables").
		Where("table_schema = ? AND table_type = ? AND table_name = ?", "public", "BASE TABLE", tableName).
		Count(&count).Error; err != nil {
		return false, err
	}
	return count > 0, nil
}

// formatValue converts a database value to a display string
func formatValue(v any) string {
	if v == nil {
		return ""
	}

	switch val := v.(type) {
	case []byte:
		return string(val)
	case string:
		return val
	default:
		return fmt.Sprintf("%v", val)
	}
}

File: rust/tealteam-web/src/handlers/db_viewer.rs
// Admin database viewer (port of Controllers/DbViewerController.cs).

use askama::Template;
use axum::extract::{Path, Query, State};
use axum::response::IntoResponse;
use axum_extra::extract::cookie::CookieJar;
use once_cell::sync::Lazy;
use sqlx::{Row, TypeInfo, ValueRef};
use std::collections::HashMap;
use tracing::error;

use super::*;
use crate::state::SharedState;

// Sensitive columns never shown in the viewer.
static SENSITIVE_COLUMNS: Lazy<HashMap<&'static str, Vec<&'static str>>> = Lazy::new(|| {
    let mut m = HashMap::new();
    m.insert("users", vec!["password_hash"]);
    m
});

pub struct TableInfo {
    pub name: String,
    pub row_count: i64,
    pub description: String,
    pub visible: bool,
    pub selected: bool,
}

pub struct ColumnInfo {
    pub name: String,
    pub type_name: String,
    pub nullable: bool,
    pub default: String,
}

#[derive(Template)]
#[template(path = "pages/db_viewer.html")]
struct DbViewerTemplate {
    title: String,
    nav: Nav,
    active_tab: String,
    tables: Vec<TableInfo>,
    selected_table: String,
}

#[derive(Template)]
#[template(path = "partials/db_table_content.html")]
struct DbTableContentFragment {
    selected_table: String,
    columns: Vec<ColumnInfo>,
    rows: Vec<Vec<String>>,
    total_rows: i64,
    offset: i64,
    limit: i64,
    shown_end: i64,
    prev_offset: i64,
    next_offset: i64,
    page: i64,
    total_pages: i64,
}

impl DbViewerTemplate {
    fn tab_class(&self, tab: &str, accent: &str) -> String {
        if self.active_tab == tab || (tab == "core" && self.active_tab.is_empty()) {
            format!("border-{accent}-500 text-{accent}-300")
        } else {
            "border-transparent text-gray-400 hover:text-gray-300".to_string()
        }
    }
}

fn tab_tables(tab: &str) -> &'static [&'static str] {
    match tab {
        "relationships" => &["event_teams"],
        "match_data" => &["matches", "scouting_data"],
        "submissions" => &["scouting_submissions", "sessions"],
        _ => &["teams", "events", "users"],
    }
}

fn table_description(name: &str, row_count: i64) -> String {
    match name {
        "scouting_data" => "Scouting data".to_string(),
        "event_teams" => "Team-event mapping".to_string(),
        "scouting_submissions" => "Pending submissions".to_string(),
        "sessions" => "User sessions".to_string(),
        _ => format!("{row_count} rows"),
    }
}

pub async fn db_viewer(
    State(state): State<SharedState>,
    jar: CookieJar,
    Query(query): Query<HashMap<String, String>>,
) -> HandlerResult {
    let Some(user) = current_user(&state.pool, &jar).await else {
        return Ok(redirect("/"));
    };
    if !user.is_admin {
        return Ok(redirect("/"));
    }

    let active_tab = query.get("tab").cloned().unwrap_or_else(|| "core".to_string());
    let selected_table = query.get("table").cloned().unwrap_or_default();
    let visible = tab_tables(&active_tab);

    let table_names: Vec<String> = sqlx::query_scalar(
        "SELECT table_name FROM information_schema.tables
         WHERE table_schema = 'public' AND table_type = 'BASE TABLE'
         ORDER BY table_name",
    )
    .fetch_all(&state.pool)
    .await
    .unwrap_or_default();

    let mut tables = Vec::new();
    for name in table_names {
        let row_count: i64 = sqlx::query_scalar(&format!("SELECT COUNT(*) FROM \"{name}\""))
            .fetch_one(&state.pool)
            .await
            .unwrap_or(0);
        let description = table_description(&name, row_count);
        let is_visible = visible.contains(&name.as_str());
        let selected = selected_table == name;
        tables.push(TableInfo { name, row_count, description, visible: is_visible, selected });
    }

    Ok(render(&DbViewerTemplate {
        title: "Database Viewer".to_string(),
        nav: Nav::from_user(Some(&user)),
        active_tab,
        tables,
        selected_table,
    }))
}

pub async fn table_content(
    State(state): State<SharedState>,
    jar: CookieJar,
    Path(name): Path<String>,
    Query(query): Query<HashMap<String, String>>,
) -> HandlerResult {
    let Some(user) = current_user(&state.pool, &jar).await else {
        return Ok(redirect("/"));
    };
    if !user.is_admin {
        return Ok(redirect("/"));
    }
    if name.is_empty() {
        return Ok((axum::http::StatusCode::BAD_REQUEST, "Table name is required").into_response());
    }

    let mut limit: i64 = query.get("limit").and_then(|v| v.parse().ok()).unwrap_or(50);
    let mut offset: i64 = query.get("offset").and_then(|v| v.parse().ok()).unwrap_or(0);
    if limit <= 0 {
        limit = 50;
    }
    if offset < 0 {
        offset = 0;
    }

    match load_table_data(&state.pool, &name, offset, limit).await {
        Ok(fragment) => Ok(render(&fragment).into_response()),
        Err(e) => {
            error!("failed to get table data for {name}: {e}");
            Ok((
                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
                format!("Failed to get table data: {e}"),
            )
                .into_response())
        }
    }
}

async fn load_table_data(
    pool: &sqlx::PgPool,
    table_name: &str,
    offset: i64,
    limit: i64,
) -> anyhow::Result<DbTableContentFragment> {
    // Validate the table name against information_schema (prevents injection).
    let allowed: i64 = sqlx::query_scalar(
        "SELECT COUNT(*) FROM information_schema.tables
         WHERE table_schema = 'public' AND table_type = 'BASE TABLE' AND table_name = $1",
    )
    .bind(table_name)
    .fetch_one(pool)
    .await?;
    if allowed == 0 {
        anyhow::bail!("invalid table name");
    }

    let all_columns: Vec<(String, String, String, Option<String>)> = sqlx::query_as(
        "SELECT column_name, data_type, is_nullable, column_default
         FROM information_schema.columns
         WHERE table_schema = 'public' AND table_name = $1
         ORDER BY ordinal_position",
    )
    .bind(table_name)
    .fetch_all(pool)
    .await?;

    let excluded: Vec<&str> = SENSITIVE_COLUMNS.get(table_name).cloned().unwrap_or_default();

    let display_columns: Vec<ColumnInfo> = all_columns
        .iter()
        .filter(|(name, _, _, _)| !excluded.contains(&name.as_str()))
        .map(|(name, type_name, nullable, default)| ColumnInfo {
            name: name.clone(),
            type_name: type_name.clone(),
            nullable: nullable == "YES",
            default: default.clone().unwrap_or_default(),
        })
        .collect();

    let total_rows: i64 = sqlx::query_scalar(&format!("SELECT COUNT(*) FROM \"{table_name}\""))
        .fetch_one(pool)
        .await?;

    let order_by = all_columns
        .first()
        .map(|(n, _, _, _)| format!(" ORDER BY \"{n}\""))
        .unwrap_or_default();
    let sql = format!("SELECT * FROM \"{table_name}\"{order_by} LIMIT {limit} OFFSET {offset}");
    let db_rows = sqlx::query(&sql).fetch_all(pool).await?;

    let mut rows = Vec::new();
    for row in &db_rows {
        let mut values = Vec::new();
        for col in &display_columns {
            values.push(format_value(row, &col.name));
        }
        rows.push(values);
    }

    let shown_end = offset + db_rows.len() as i64;
    Ok(DbTableContentFragment {
        selected_table: table_name.to_string(),
        columns: display_columns,
        rows,
        total_rows,
        offset,
        limit,
        shown_end,
        prev_offset: (offset - limit).max(0),
        next_offset: offset + limit,
        page: offset / limit + 1,
        total_pages: total_rows / limit + 1,
    })
}

/// Best-effort stringification of an arbitrary column value for display.
fn format_value(row: &sqlx::postgres::PgRow, col_name: &str) -> String {
    let Ok(value) = row.try_get_raw(col_name) else {
        return String::new();
    };
    if value.is_null() {
        return String::new();
    }

    let type_info = value.type_info();
    let type_name = type_info.name();
    // Try the common types in turn; fall back to the type name if decoding fails.
    match type_name {
        "INT2" => row.try_get::<i16, _>(col_name).map(|v| v.to_string()).unwrap_or_default(),
        "INT4" => row.try_get::<i32, _>(col_name).map(|v| v.to_string()).unwrap_or_default(),
        "INT8" => row.try_get::<i64, _>(col_name).map(|v| v.to_string()).unwrap_or_default(),
        "FLOAT4" => row.try_get::<f32, _>(col_name).map(|v| v.to_string()).unwrap_or_default(),
        "FLOAT8" => row.try_get::<f64, _>(col_name).map(|v| v.to_string()).unwrap_or_default(),
        "NUMERIC" => row
            .try_get::<f64, _>(col_name)
            .map(|v| v.to_string())
            .or_else(|_| row.try_get::<i64, _>(col_name).map(|v| v.to_string()))
            .unwrap_or_default(),
        "BOOL" => row.try_get::<bool, _>(col_name).map(|v| v.to_string()).unwrap_or_default(),
        "TIMESTAMPTZ" => row
            .try_get::<chrono::DateTime<chrono::Utc>, _>(col_name)
            .map(|v| v.to_rfc3339())
            .unwrap_or_default(),
        "TIMESTAMP" => row
            .try_get::<chrono::NaiveDateTime, _>(col_name)
            .map(|v| v.to_string())
            .unwrap_or_default(),
        "DATE" => row
            .try_get::<chrono::NaiveDate, _>(col_name)
            .map(|v| v.to_string())
            .unwrap_or_default(),
        _ => row.try_get::<String, _>(col_name).unwrap_or_else(|_| String::new()),
    }
}
